add endpoint for multipart data upload

xichen-ansearch opened the pull request:
Hi,

This PR adds support for file uploads via a multipart endpoint. It allows users to upload files along with their associated documentID, which is then processed and stored for further use. There are a couple of points I’d like to clarify regarding this implementation:

**1. Document ID Assignment:**
Currently, this implementation allows uploading one file at a time, with a documentID provided for each file. However, if we want to support uploading multiple files in a single request, there's no clear way to assign a unique documentID to each file. I’d like to confirm: should we maintain the current approach where users manually provide a documentID for each file?

**2. File Storage Location:**
Regarding file uploads, Fastify requires that files be stored in a specific way (e.g., using temporary directories or streams). In the current implementation, files are directly processed as streams in memory, and no files are written to disk. Would it be preferable to store uploaded files in a temporary location before processing, or is the current memory-based approach acceptable?

Thanks. This PR will fix #73

Mael-Abgrall commented on src/api/search/api-search-routes.ts:
```diff
@@ -257,4 +259,84 @@ Note:
+          'Upload multiple files with a required document-id header.',
+        response: { 200: uploadFileResponseSchema },
+        security: [{ bearerAuth: [] }],
+      },
```
A few comments and changes you need to do:

- First of all respect standards. Headers are for the identification of requests and adding parameters that are read by servers, Ex: API tokens, server IPs, referral domains. This not for content (you should not send document ID or user data in there). [Read more about this here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers)
- Second, please leverage the body. It will allow to have KV pairs with the document name and the ID the user wants to attribute to it (meaning you can upload multiple documents, and give them multiple IDs).
- Finally, you may not see it, but there are protection built-in with Fastify, we are using schemas for this exact reason (whenever you see a typebox schema like Type.Object, this is a schema); This mean that anyone calling the API with the wrong body will be handled for you (eg: you are expecting  a parameter called ID, the user does not send this ID, the request will be automatically rejected). This mean the code line 296 to 301 would be automatically done for you (when using body or parameters, this system does not work with headers, again because headers are handled differently).

Mael-Abgrall replied:
Additionally, your header is not documented, which is bad practice (our users won't know you need to add this header)

Mael-Abgrall commented on src/api/search/api-search-routes.ts:
```diff
@@ -257,4 +259,84 @@ Note:
+
+  app.register(multipart);
+  app.post(
+    '/upload/files',
```
The path will need to change.

I would like to have all three ingest endpoints to start with ingest.
```
/search/ingest/raw
/search/ingest/urls
/search/ingest/multipart
```

Would you mind editing the other endpoints as well?

Mael-Abgrall commented on src/api/search/api-search-routes.ts:
```diff
@@ -257,4 +259,84 @@ Note:
+
+            await secureVerifyDocumentID({ documentID });
+
+            const content = await parseStream({
```
Normally, because we are using an hexagonal architecture, you should not call parseStream here (you are in a function of the hexagon, and the only things you can call are the core)

I'll leave it here for now, as I'd like to create something simple, and add complexity and resiliency later (once this endpoint is correctly made)

Mael-Abgrall commented on src/api/search/api-search-routes.ts:
```diff
@@ -257,4 +259,84 @@ Note:
+              });
+            }
+
+            await secureVerifyDocumentID({ documentID });
```
do this before processing documents. It's a lot faster to check every single id than downloading one document.

Doing so help also our users realise something is wrong much sooner (eg: wait 3s to have a rejection vs wait 10ms to have a rejection) leading to better cost both for them and us

Mael-Abgrall reviewed (changes requested):
Okay, you're on a good start, but more work is needed.

xichen-ansearch commented:
Hi, here are a few updates I've made:

1. moved multipart registration to the server level
2. documented the multipart headers for clarity
3. updated the body handling to allow users to upload multiple documents with unique IDs
4. reorganized validations like documentID checks to run before file processing for better efficiency.

Thanks!

Mael-Abgrall commented on src/api/ingest/api-ingest-routes.ts:
```diff
@@ -182,4 +183,170 @@ Ideally your text should be in markdown format, to help with context extraction.
+        description: `
+Upload multiple files and specify a unique documentID for each file.
+
+**Form field 1**: documents - A JSON array describing the metadata for each file, including documentID. For example:
```
To make your doc cleared, instead of saying 'form field 1'; present it with form fields:

```
The form has two fields, one for the files metadata, and one for the files themselves.

- 'documents' is the field for adding metadata; it's a JSON array etc
- 'files' (multiple files) - Each file corresponds to an entry in the documents array.

the number of items in the documents array must equal the number of files uploaded, or the API will return an error.
```

Mael-Abgrall replied:
Additionally, please add an 's' to the `file` field

Mael-Abgrall reviewed (changes requested):
Good work, please address the comment on the documentation and the field name

xichen-ansearch commented:
Hi,

I've updated the description.

Mael-Abgrall reviewed (approved):
Good job :)

Now there's an ultimate task for you, update the website doc.
The file is here:
https://github.com/A-star-logic/a-star-logic/blob/main/frontends/apps/astarlogic.com/src/content/doc/1.Memoire/2.Send%20and%20search%20documents.md
